Modify meta doc (#1230)
php-fig/fig-standards: proposed/clock-meta.md

## 1. Summary

> Getting the current time in applications is a rather straightforward thing given the language's native `time()` function (or one of its similar functionalities like `utime()` or `new DateTimeImmutable()` etc.)

Reviewer: s/utime/microtime I suppose?

Author: Good catch! Thanks. Mixed up the underlying C-functions there 😁

> Getting the current time in applications is a rather straight forward thing given the languages native `time()` function (or one of it's similar functionlaities like `utime()` or `new DateTimeImmutable()` etc.)

Reviewer 2: ```suggestion
Getting the current time in applications is a rather straight forward thing given the languages native `time()` function (or one of it's similar functionalities like `utime()` or `new DateTimeImmutable()` etc.)
```

> Getting the current time in applications is a rather straight forward thing given the languages native `time()` function (or one of it's similar functionlaities like `utime()` or `new DateTimeImmutable()` etc.)

Reviewer 3: s/straight forward/straightforward/ (it's one word)

s/given the languages native/given the language's native/ (make it possessive)

s/one of it's similar/one of its similar/ (make it possessive rather than "it is" contraction)

Author: Fixed with f5c2188

> Getting the current time in applications is a rather straightforward thing given the language's native `time()` function (or one of its similar functionalities like `microtime()` or `new DateTimeImmutable()` etc.)

Reviewer 4: ```suggestion
Getting the current time in applications is typically achieved using the `time()` or `microtime` functions, or by using a `new \DateTimeImmutable()` class.
```

> Due to the nature of time as constantly progressing it becomes a nuisance to use these functions if you need predictable results like for testing.

Reviewer 2: ```suggestion
Due to the nature of time as constantly progressing it becomes a nuisance to use these functions if you need predictable results, for example when testing.
```

> Due to the nature of time as constantly progressing it becomes a nuisance to use these functions if you need predictable results, for example when testing.

Reviewer 4: ```suggestion
Due to the nature of time progression these methods cannot be used when predictable results are needed, such as during testing.
```

> This `ClockInterface` aims to provide a standard way to consume time, that allows interopability not only when consuming the "real" time but also when predictable results need to be available. This allows to skip using PHP Extensions for testing or using hacks like re-declaring the `time()` function in other namespaces. 

Reviewer 2: ```suggestion
This `ClockInterface` aims to provide a standard way to consume time, that allows interoperability not only when consuming the "real" time but also when predictable results need to be available. This allows the avoidance of using PHP Extensions for testing or using hacks like re-declaring the `time()` function in other namespaces. 
```

> This `ClockInterface` aims to provide a standard way to consume time, that allows interoperability not only when consuming the "real" time but also when predictable results need to be available. This allows the avoidance of using PHP Extensions for testing or using hacks like re-declaring the `time()` function in other namespaces. 

Author: In that context rewriting a function in a namespace *is* a hack ;-)

But I do get your point :wink:

## 2. Why Bother?

There are currently a few libraries that provide this functionality, however there is no interopability between these different libraries, as they ship with their own clock interfaces. 

> Symfony provides a package called `symfony/phpunit-bridge` that has a
> `Symfony\Bridge\PhpUnit\ClockMock` class, which allows mocking PHP's built-in time & date functions, however this does not solve mocking calls to `new \DateTimeImmutable()`. It does not fully mock time when called from other libraries that rely on the system time. 
> `Cake\Chronos\Chronos` does provide mocking, however it is set via a global (static class property), and this has its own pitfalls as it provides no isolation.

Reviewer 4: ```suggestion
Symfony provides a package called `symfony/phpunit-bridge` that has a `Symfony\Bridge\PhpUnit\ClockMock` class, which allows mocking PHP's built-in time and date functions, however this does not solve mocking calls to `new \DateTimeImmutable()`. It also does not fully mock time when called from other libraries that rely on the system time.

`Cake\Chronos\Chronos` does provide mocking via a static `setTestNow()` method, but this has its own pitfalls as it provides no isolation and must be called again to stop mocking.
```

> `Cake\Chronos\Chronos` does provide mocking via a static `setTestNow()` method, but this has its own pitfalls as it provides no isolation and must be called again to stop mocking.

Reviewer 5: `Chronos` just inherits `setTestNow()` from `Carbon`, the project they forked, which implemented it years before. By respect for @briannesbitt, the original author, please mention `Carbon` instead.

Author: This should now be fixed by mentioning Carboin as well as Chronos

  described in this document, so it is not a coding standard;
* This PSR does not provide a recommendation on how to handle timezones when 
  retrieving the current time. This is left up to the implementation.
* This PSR does not handle any scheduling methods like `sleep()` or `wait()` because such methods are not related to retrieving the current time.

## 4. Approaches

### 4.1 Chosen Approach

We have decided to formalize the existing practices, used by several other packages out in the wild. Some of the popular packages providing this functionality are: 

* [`lcobucci/clock`](https://packagist.org/packages/lcobucci/clock)
* [`kreait/clock`](https://packagist.org/packages/kreait/clock)
* [`ergebnis/clock`](https://packagist.org/packages/ergebnis/clock)
* [`mangoweb/clock`](https://packagist.org/packages/mangoweb/clock)

(This list is not exhaustive!)

> interfaces, and some relying on overloading (extending) the Clock class to mock the

Reviewer 5: "Some providing" got truncated.
```suggestion
Some providing interfaces, and some relying on overloading (extending) the Clock class to mock the
```

Author: Thanks! Fixed that

> 

Author: Even though that should not be necessary I'd rather keep it to make clear that there might be other packages that provide the same value. Yes, we introduced that list with "some popular packages … are". But there will be people not reading that. And the oneliner right underneath the table is catching the eye better than the message in the copy-text above.

> Some of these providing interfaces, and some relying on overloading (extending) the Clock class to mock the

Reviewer 4: ```suggestion
Some of these provide interfaces and some rely on extending a clock class to mock the
```

current time.

These implementations all provide a `now()` method which returns a `DateTimeImmutable` object. As the `DateTimeImmutable` object allows retrieving the Unix timestamp, by calling `getTimestamp()` or `format('u.U')`, this interface does not define any special methods to retrieve a Unix timestamp or any other time information that is not available from a `DateTimeImmutable` object. 

### 4.2 Timezones

Time by now is defined by interaction of electromagnetic radiation with the excited states of certain atoms where the SI defines one second as the duration of 9192631770 cycles of radiation corresponding to the transition between two energy levels of the ground state of the caesium-133 atom at 0K. This means that retrieving the current time will always return the same time, no matter where it is observed. While the timezone defines *where* the time was observed it does not modify the actual "slice" of time.

This means that for the sake of this PSR the timezone is considered an implementation detail of the interface. 

> It is up to the implementation to make sure that the timezone is handled according to the business logic of the application. That is either by making sure that a call to `now()` will only return a `DateTimeImmutable` object with a known timezone (implicit contract) or by explicitly changing the timezone to be correct for the application. This an be done by calling `setTimezone()` to create a new `DateTimeImmutable` object with the given timezone. 

Reviewer 5: Yes x 1000! This paragraph is the most important to me. I hope the one above about caesium-133 atom won't loose readers before they reach this 🤣

Author: Good catch! Thanks for spotting!

> These are though actions that are not part of the contract which this interface defines.

Reviewer 4: ```suggestion
These actions are not defined in this interface.
```

  described in this document, so it is not a coding standard;
* This PSR does not provide a recommendation on how to handle timezones when 
  retrieving the current time. This is left up to the implementation.

> * This PSR does not handle any Scheduler specific tasks like `sleep` or `wait` as those – even though they take a time or an interval as parameter – are not related to retrieving the current time. Handling such tasks is left to the respective implementations or a different PSR.

Reviewer 4: ```suggestion
* This PSR does not handle any scheduling methods like `sleep()` or `wait()` because such methods are not related to retrieving the current time.
```

## 4. Approaches

### 4.1 Chosen Approach

We have decided to formalize the existing practices, used by several other packages out in the wild. Some of the popular packages providing this functionality are: 

* [`lcobucci/clock`](https://packagist.org/packages/lcobucci/clock)
* [`kreait/clock`](https://packagist.org/packages/kreait/clock)
* [`ergebnis/clock`](https://packagist.org/packages/ergebnis/clock)
* [`mangoweb/clock`](https://packagist.org/packages/mangoweb/clock)

(This list is not exhaustive!)

> Some providing

Reviewer 2: ```suggestion
```

interfaces, and some relying on overloading (extending) the Clock class to mock the
current time.

> These implementations all provide a `now`-method which returns a `DateTimeImmutable`-object. As the `DateTimeImmutable`-object allows to retieve the Unix-Timestamp by calling `getTimestamp()` or `format('u.U')` this interface does not define any special methods to retieve a Unix-Timestamp or any other form of time-information apart from retrieving the `DateTimeImmutable`-object. 

Reviewer 4: ```suggestion
These implementations all provide a `now()` method which returns a `DateTimeImmutable` object. As the `DateTimeImmutable` object allows retrieving the Unix timestamp, by calling `getTimestamp()` or `format('u.U')`, this interface does not define any special methods to retrieve a Unix timestamp or any other time information that is not available from a `DateTimeImmutable` object. 
```

### 4.2 Timezones

Time by now is defined by interaction of electromagnetic radiation with the excited states of certain atoms where the SI defines one second as the duration of 9192631770 cycles of radiation corresponding to the transition between two energy levels of the ground state of the caesium-133 atom at 0K. This means that retrieving the current time will always return the same time, no matter where it is observed. While the timezone defines *where* the time was observed it does not modify the actual "slice" of time.

> This means that for the sake of this PSR the timezone is irrelevant as it is an implementation detail of the code implementing the interface. 

Reviewer 4: ```suggestion
This means that for the sake of this PSR the timezone is considered an implementation detail of the interface. 
```

> As there is no way to assert a certain timezone in the returnd `DateTimeImmutable` it is up to the implementation to make sure that the timezone is handled according to the Business-Logic. That is either by making sure that a call to `now()` will only return a `DateTimeImmutable`-object with a known timezone (implicit contract) or by making sure that the Business-logic handles the timezone explicitly. This an be done by calling `setTimezone` or by using `getTimestamp` which will automatically return the UTC based Unix-Timestamp.

Reviewer 4: ```suggestion
It is up to the implementation to make sure that the timezone is handled according to the business logic of the application. That is either by making sure that a call to `now()` will only return a `DateTimeImmutable` object with a known timezone (implicit contract) or by explicitly changing the timezone to be correct for the application. This an be done by calling `setTimezone()` to create a new `DateTimeImmutable` object with the given timezone. 
```

> 
> These are though actions that are not part of the contract which this interface defines.

Reviewer 4: ```suggestion
```

Author: We should make it explicit that handling the timezones are not part of the interfaces contract!

    }
}

//

final class UTCClock implements \Psr\Clock\ClockInterface
{
    private \DateTimeZone $utcTimeZone;

    public function __construct()
    {
        $this->utcTimeZone = new \DateTimeZone('UTC');
    }

    public function now(): \DateTimeImmutable
    {
        return new \DateTimeImmutable('now', $this->utcTimeZone);
    }

> }

Reviewer 4: Why was this removed?

Author: If we are proposing that the timezone is not part of the Interface we should not provide a minimum implementation that provides a timezone specific implementation. And for a minimum implementation the UTCClock is not necessary. If people want a more sophisticated example they can check out one of the multiple provided implementations
